File: typescript/src/middleware/audit.ts
import type { Decision } from "./types.js";
import { Action } from "./types.js";

export interface AuditRecord {
  session_id: string;
  tool_name: string;
  args: Record<string, unknown>;
  decision: Decision;
  timestamp: Date;
}

export interface AuditFilter {
  session_id?: string;
  tool_name?: string;
  action?: Action;
  since?: Date;
  limit?: number;
}

export interface Auditor {
  record(record: AuditRecord): void;
  query(filter: AuditFilter): AuditRecord[];
}

export class InMemoryAuditor implements Auditor {
  private records: AuditRecord[] = [];

  record(record: AuditRecord): void {
    this.records.push(record);
  }

  query(filter: AuditFilter): AuditRecord[] {
    const results: AuditRecord[] = [];
    for (const r of this.records) {
      if (filter.session_id && r.session_id !== filter.session_id) continue;
      if (filter.tool_name && r.tool_name !== filter.tool_name) continue;
      if (filter.action && r.decision.action !== filter.action) continue;
      if (filter.since && r.timestamp < filter.since) continue;
      results.push(r);
      if (filter.limit && results.length >= filter.limit) break;
    }
    return results;
  }

  clear(): void {
    this.records = [];
  }
}

File: go/middleware/audit.go
package middleware

import "time"

type AuditRecord struct {
	SessionID string
	ToolName  string
	Args      map[string]any
	Decision  Decision
	Timestamp time.Time
}

type Auditor interface {
	Record(record AuditRecord)
	Query(filter AuditFilter) []AuditRecord
}

type AuditFilter struct {
	SessionID string
	ToolName  string
	Action    Action
	Since     time.Time
	Limit     int
}

type InMemoryAuditor struct {
	records []AuditRecord
}

func NewInMemoryAuditor() *InMemoryAuditor {
	return &InMemoryAuditor{
		records: make([]AuditRecord, 0),
	}
}

func (a *InMemoryAuditor) Record(record AuditRecord) {
	a.records = append(a.records, record)
}

func (a *InMemoryAuditor) Query(filter AuditFilter) []AuditRecord {
	result := make([]AuditRecord, 0)
	for _, r := range a.records {
		if filter.SessionID != "" && r.SessionID != filter.SessionID {
			continue
		}
		if filter.ToolName != "" && r.ToolName != filter.ToolName {
			continue
		}
		if filter.Action != "" && r.Decision.Action != filter.Action {
			continue
		}
		if !filter.Since.IsZero() && r.Timestamp.Before(filter.Since) {
			continue
		}
		result = append(result, r)
		if filter.Limit > 0 && len(result) >= filter.Limit {
			break
		}
	}
	return result
}
